cdap-tms/src/main/java/co/cask/cdap/messaging/subscriber/AbstractMessagingPollingService.java

```diff
@@ -0,0 +1,267 @@
+   */
+  protected void postProcess() {
+    // no-op
+  }
+
+  @Override
+  protected final boolean shouldRetry(Exception ex) {
+    // Log the exception
+    try {
+      throw ex;
+    } catch (ServiceUnavailableException e) {
+      SAMPLING_LOG.warn("Failed to contact service {}. Will retry in next run.", e.getServiceName(), e);
+    } catch (TopicNotFoundException e) {
+      SAMPLING_LOG.warn("Failed to fetch from TMS. Will retry in next run.", e);
+    } catch (Exception e) {
+      SAMPLING_LOG.warn("Failed to get and process notifications. Will retry in next run", e);
+    }
+
+    return true;
+  }
+
+  @Override
+  protected final long runTask() throws Exception {
+    long delayMillis = fetchAndProcessMessages();
+    try {
+      postProcess();
+    } catch (Exception e) {
+      LOG.warn("Failed to perform post processing after processing messages.", e);
+    }
+    return delayMillis;
+  }
+
+  /**
+   * Fetching messages from TMS.
+   */
+  protected List<Message> fetchMessages(@Nullable String messageId) throws TopicNotFoundException, IOException {
+    List<Message> messages = new ArrayList<>();
+    LOG.trace("Fetching from topic '{}' with messageId '{}'", topicId, messageId);
+    MessageFetcher messageFetcher = getMessagingContext().getMessageFetcher();
+    try (CloseableIterator<Message> iterator = messageFetcher.fetch(getTopicId().getNamespace(),
+                                                                    getTopicId().getTopic(), fetchSize, messageId)) {
+      while (iterator.hasNext() && state() == State.RUNNING) {
+        messages.add(iterator.next());
+      }
+    }
+    return messages;
+  }
+
+  /**
+   * The method has the main logic to perform one fetch from TMS and process the fetched messages.
+   *
+   * @return number of milliseconds to sleep before the next fetch and process should happen.
+   */
+  private long fetchAndProcessMessages() throws TopicNotFoundException, IOException {
+    // Fetch the messageId if hasn't been fetched
+    if (!messageIdInitialized) {
+      messageId = loadMessageId();
+      messageIdInitialized = true;
+    }
+
+    long startTime = System.currentTimeMillis();
+
+    // Collects batch of messages for processing.
+    List<Message> messages = fetchMessages(messageId);
+    metricsContext.gauge("tms.fetch.time.ms", System.currentTimeMillis() - startTime);
+    metricsContext.increment("tms.fetch.messages", messages.size());
+
+    // Return if stopping or request to sleep for configured number of milliseconds if there are no notifications
+    if (messages.isEmpty() || state() != State.RUNNING) {
+      return emptyFetchDelayMillis;
+    }
+
+    startTime = System.currentTimeMillis();
+
+    MessageIterator iterator = new MessageIterator(messages.iterator());
+    try {
+      String messageId = processMessages(iterator);
+      this.messageId = messageId == null ? this.messageId : messageId;
+    } catch (TopicNotFoundException | IOException e) {
```

Comment:
Hum... since the `runTask` method has `throws Exception`, actually we don't need to catch here at all. Updating the method signature to `throws Exception`.